From 43c34236c7e02b924ddd430b57beb58fada89011 Mon Sep 17 00:00:00 2001
From: Mark Rushakoff <mark@strange.love>
Date: Wed, 17 Jul 2024 10:49:18 -0400
Subject: [PATCH 3/4] fix: respect command context in server start

---
 server/v2/commands.go | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/server/v2/commands.go b/server/v2/commands.go
index 314e2c5f6e..938f7cd1a1 100644
--- a/server/v2/commands.go
+++ b/server/v2/commands.go
@@ -121,9 +121,13 @@ func createStartCommand[T transaction.Tx](
 			go func() {
 				sigCh := make(chan os.Signal, 1)
 				signal.Notify(sigCh, syscall.SIGINT, syscall.SIGTERM)
-				sig := <-sigCh
-				cancelFn()
-				cmd.Printf("caught %s signal\n", sig.String())
+				select {
+				case sig := <-sigCh:
+					cancelFn()
+					cmd.Printf("caught %s signal\n", sig.String())
+				case <-ctx.Done():
+					cancelFn()
+				}
 
 				if err := server.Stop(ctx); err != nil {
 					cmd.PrintErrln("failed to stop servers:", err)
-- 
2.44.0

